diff --git a/node_modules/expo-sqlite/web/WorkerChannel.ts b/node_modules/expo-sqlite/web/WorkerChannel.ts
index 7fba380..80b8af9 100644
--- a/node_modules/expo-sqlite/web/WorkerChannel.ts
+++ b/node_modules/expo-sqlite/web/WorkerChannel.ts
@@ -40,7 +40,7 @@ export function sendWorkerResult({
     const resultJson = error != null ? serialize({ error }) : serialize({ result });
     const resultBytes = new TextEncoder().encode(resultJson);
     const length = resultBytes.length;
-    resultArray.set(new Uint32Array([length]), 0);
+    new Uint32Array(resultBuffer, 0, 1)[0] = length;
     resultArray.set(resultBytes, 4);
     Atomics.store(lock, 0, RESOLVED);
   } else {
@@ -117,23 +117,16 @@ export function invokeWorkerSync<T extends SQLiteWorkerMessageType & keyof Resul
   });
 
   let i = 0;
+  const deadline = performance.now() + 5000;
   // @ts-expect-error: Remove this when TypeScript supports Atomics.pause
   const useAtomicsPause = typeof Atomics.pause === 'function';
   while (Atomics.load(lock, 0) === PENDING) {
-    ++i;
-
+    if ((++i % 1024) === 0 && performance.now() >= deadline) {
+      throw new Error('Sync operation timeout');
+    }
     if (useAtomicsPause) {
-      if (i > 1_000_000) {
-        throw new Error('Sync operation timeout');
-      }
       // @ts-expect-error: Remove this when TypeScript supports Atomics.pause
       Atomics.pause();
-    } else {
-      // NOTE(kudo): Unfortunate for the busy loop,
-      // because we don't have a way for main thread to yield its execution to other callbacks.
-      if (i > 1000_000_000) {
-        throw new Error('Sync operation timeout');
-      }
     }
   }
 
